Failing to build libuavcan

When trying to build by following the build instructions in the README, or when running:

```
set -e
cd /tmp
git clone https://github.com/UAVCAN/libuavcan -b v1.0.0rc5 --recurse-submodules
cd libuavcan
# Unset flags in case my defaults are interfering
unset CFLAGS CXXFLAGS CPPFLAGS LDFLAGS MAKEFLAGS
# Set explicitly, to avoid using hardening wrapper
export PATH=/usr/bin
mkdir -p build
cd build
cmake ..
# Disable parallel build in case that's causing problems
make -j1
```

I see a load of warnings regarding `-Wexpansion-to-defined` for ifdef's in various headers, followed by fatal errors.

# Tail of build output including various warnings and the fatal error:

```
 #if UAVCAN_EXCEPTIONS
     ^~~~~~~~~~~~~~~~~
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
/tmp/libuavcan/./libuavcan/include/uavcan/protocol/logger.hpp:274:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
In file included from /tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/uavcan_linux.hpp:10:0,
                 from /tmp/libuavcan/libuavcan_drivers/linux/apps/test_posix.cpp:7:
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp: In static member function ‘static can_frame uavcan_linux::SocketCanIface::makeSocketCanFrame(const uavcan::CanFrame&)’:
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:64:105: warning: missing initializer for member ‘can_frame::__res0’ [-Wmissing-field-initializers]
         ::can_frame sockcan_frame { uavcan_frame.id& uavcan::CanFrame::MaskExtID, uavcan_frame.dlc, { } };
                                                                                                         ^
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:64:105: warning: missing initializer for member ‘can_frame::__res1’ [-Wmissing-field-initializers]
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:64:105: warning: missing initializer for member ‘can_frame::data’ [-Wmissing-field-initializers]
In file included from /usr/include/features.h:423:0,
                 from /usr/include/c++/7.2.0/x86_64-pc-linux-gnu/bits/os_defines.h:39,
                 from /usr/include/c++/7.2.0/x86_64-pc-linux-gnu/bits/c++config.h:533,
                 from /usr/include/c++/7.2.0/string:38,
                 from /tmp/libuavcan/./libuavcan/include/uavcan/build_config.hpp:138,
                 from /tmp/libuavcan/./libuavcan/include/uavcan/protocol/dynamic_node_id_server/event.hpp:8,
                 from /tmp/libuavcan/libuavcan_drivers/posix/include/uavcan_posix/dynamic_node_id_server/file_event_tracer.hpp:12,
                 from /tmp/libuavcan/libuavcan_drivers/linux/apps/test_posix.cpp:5:
/usr/include/bits/socket.h: In member function ‘int uavcan_linux::SocketCanIface::read(uavcan::CanFrame&, uavcan::UtcTime&, bool&) const’:
/usr/include/bits/socket.h:279:45: error: flexible array member ‘cmsghdr::__cmsg_data’ not at end of ‘struct uavcan_linux::SocketCanIface::read(uavcan::CanFrame&, uavcan::UtcTime&, bool&) const::Control’
     __extension__ unsigned char __cmsg_data __flexarr; /* Ancillary data.  */
                                             ^
In file included from /tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/uavcan_linux.hpp:10:0,
                 from /tmp/libuavcan/libuavcan_drivers/linux/apps/test_posix.cpp:7:
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:225:48: note: next member ‘uint8_t uavcan_linux::SocketCanIface::read(uavcan::CanFrame&, uavcan::UtcTime&, bool&) const::Control::data [16]’ declared here
             std::uint8_t data[sizeof(::timeval)];
                                                ^
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:222:16: note: in the definition of ‘struct uavcan_linux::SocketCanIface::read(uavcan::CanFrame&, uavcan::UtcTime&, bool&) const::Control’
         struct Control
                ^~~~~~~
/tmp/libuavcan/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp:227:32: error: value-initialization of incomplete type ‘unsigned char []’
         auto control = Control();
                                ^
make[2]: *** [libuavcan_drivers/linux/CMakeFiles/test_posix.dir/build.make:63: libuavcan_drivers/linux/CMakeFiles/test_posix.dir/apps/test_posix.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:214: libuavcan_drivers/linux/CMakeFiles/test_posix.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

# Cmake output:

```
-- The C compiler identification is GNU 7.2.0
-- The CXX compiler identification is GNU 7.2.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using C++11 (pass UAVCAN_USE_CPP03=1 to override)
-- Release build type: RelWithDebInfo
-- Adding UAVCAN Linux platform driver
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Using uavcan target; source dir: /tmp/libuavcan/libuavcan
-- POSIX source dir: /tmp/libuavcan/libuavcan_drivers/posix
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/libuavcan/build
```

# Tool versions:

```
$ gcc --version
gcc (GCC) 7.2.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ cmake --version
cmake version 3.9.6

CMake suite maintained and supported by Kitware (kitware.com/cmake).

$ lsb_release -a
LSB Version:    1.4
Distributor ID: Arch
Description:    Arch Linux
Release:        rolling
Codename:       n/a

$ uname -a
Linux hai 4.14.0-mark #1 SMP PREEMPT Mon Nov 13 16:50:57 GMT 2017 x86_64 GNU/Linux
```

--- a/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp
+++ b/libuavcan_drivers/linux/include/uavcan_linux/socketcan.hpp
@@ -219,18 +219,17 @@ class SocketCanIface : public uavcan::ICanIface
         iov.iov_base = &sockcan_frame;
         iov.iov_len  = sizeof(sockcan_frame);
 
-        struct Control
-        {
-            cmsghdr cm;
-            std::uint8_t data[sizeof(::timeval)];
-        };
-        auto control = Control();
+        static constexpr size_t ControlSize = sizeof(cmsghdr) + sizeof(::timeval);
+        using ControlStorage = typename std::aligned_storage<ControlSize>::type;
+        ControlStorage control_storage;
+        auto control = reinterpret_cast<std::uint8_t *>(&control_storage);
+        std::fill(control, control + ControlSize, 0x00);
 
         auto msg = ::msghdr();
         msg.msg_iov    = &iov;
         msg.msg_iovlen = 1;
-        msg.msg_control = &control;
-        msg.msg_controllen = sizeof(control);
+        msg.msg_control = control;
+        msg.msg_controllen = ControlSize;
 
         const int res = ::recvmsg(fd_, &msg, MSG_DONTWAIT);
         if (res <= 0)

